JavaScriptCodeConventions.md

```diff
@@ -0,0 +1,70 @@
+# JavaScript Code Conventions
+> NOTE: These conventions also apply to super-sets of JavaScript which are used within Ultimaker such as TypeScript.
+
+## Introduction
+These JavaScript style guidelines are intended to be used by both the R&D and marketing team as part of a broader effort to standardize JavaScript style across the company and as an early facilitator for the fact that Ultimaker continues to offer more web-based products such as the Cura Cloud services and Cura Connect.
+
+## AirBnB as Baseline Style
+To reduce the need to build comprehensive style guidelines ourselves, we have decided to adopt the well-known and widely used **AirBnB style**. The full documentation for AirBnB style [can be found on GitHub](https://github.com/airbnb/javascript). For the most part, this style is already very similar to what is used by front-end teams at Ultimaker with a few differences, which are outlined below. It was also chosen because of it was designed for usage with React.js, which is used by the R&D department. It is applicable to Vue.js as well, however, and can be used by the marketing department.
+
+## Enforcing AirBnB Style in Your Editor
+### Set-Up for VS Code
+1. Install packages:
+
+```
+npm i --save-dev eslint eslint-config-airbnb eslint-plugin-import eslint-plugin-jsx-a11y eslint-plugin-react typescript-eslint-parser
+```
+2. Install the [`eslint` plugin for VS Code](https://marketplace.visualstudio.com/items?itemName=dbaeumer.vscode-eslint).
+3. Create an `.eslintrc.json` file in your project’s root directory.
+4. Add the following content to `.eslintrc.json`:
+
+```
+{
+    "extends": "airbnb",
+    "env": {
+        "node": true,
+        "es6": true,
+        "browser": true
+    },
+    "parser": "typescript-eslint-parser",
+    "parserOptions": {
+        "ecmaVersion": 6,
+        "sourceType": "module",
+        "ecmaFeatures": {
+            "modules": true,
+            "jsx": true
+        }
+    },
+    "rules": {
+        "indent": ["error", 4],
+        "no-underscore-dangle": 0
+    },
+    "overrides": [
+        {
+            "files": [
+                "**/*.ts",
+                "**/*.tsx"
+            ]
+        }
+    ]
+}
+```
+
+## Additions to AirBnB
+### Comments
+Use JSDoc style comments.
+
+### Typing
+TypeScript types/interfaces should use PascalCase.
+
+## Exceptions to AirBnB
+### Tabs
+Although AirBnB prescribes 2 spaces for indentation, _all_ of Ultimaker software (R&D and Marketing) already use 4 spaces as a standard. For this reason an exception is added in `.eslintrc.json`.
+
+### Underscores
+One difficult change which will need to be adapted over time is the elimination of underscores. For this rule, an exception can be made until projects are converted to TypeScript, at which point the AirBnB style will be followed with an additional caveat.
```

Comment:
Not all Javascript engines in use are able to support TypeScript, so this will be forever.

Reply:
I'm not sure I understand. No JS engine supports TS, TS is always compiled to JS.

Reply:
Let me put it this way then: No TypeScript compiler is going to be able to compile TypeScript embedded in QML and/or insert JavaScript back into QML.

Reply:
That's true although we never use classes and private/protected variables in QML anyway. The argument for the rule has nothing to do with TypeScript in fact: the rationale is that since they're not really private/protected, making fake private/protected is a hack. In contexts without real private variables (i.e. not using TypeScript's `private` or block-scoped `let` and `const`, you should assume the variable is public... because it is).

Reply:
I don't see underscores as a hack to make them 'private', it's a naming convention so it's easy to see what the author indented with the method and if you can change the signature without breaking any potential dependencies on the method.

Reply:
Let's just go with the exception of allowing underscores, but lets say that in typescript we recommend using the `private` key word for private functions and `const` and `let` for private variables.